qiskit_experiments/analysis/curve_fitting.py

```diff
@@ -0,0 +1,220 @@
+):
+    """Use non-linear least squares to fit a list of functions, f_i, to data
+
+    Args:
+        funcs: a list of objective functions with signatures `f_i(x, *params)`.
+        xdata: a 2D array of xdata and function function indexes.
+        ydata: a 1D array of ydata
+        p0: initial guess for optimization parameters.
+        sigma: Optional, a 1D array of standard deviations in ydata.
+        weights: Optional, a 1D list of numeric weights for each function.
+        kwargs: additional kwargs for scipy.optimize.curve_fit.
+
+    Returns:
+        AnalysisResult: result containing `popt` the optimal fit parameters,
+                        `popt_err` the standard error estimates popt,
+                        `pcov` the covariance matrix for the fit,
+                        `chisq` the chi-squared parameter of fit.
+                        `xrange` the range of xdata values used for fit.
+
+    Raises:
+        QiskitError: if input xdata is not 2D.
+    """
+    num_funcs = len(funcs)
+
+    # Get 1D xdata and indices from 2D input xdata
+    xdata = np.asarray(xdata, dtype=float)
+    if xdata.ndim != 2:
+        raise QiskitError("multi_curve_fit requires 2D xdata.")
+    xdata1d, xindex = xdata.T
+
+    # Get positions for indexes data sets
+    idxs = [xindex == i for i in range(num_funcs)]
+
+    # Combine weights and sigma for transformation
+    if weights is None:
+        wsigma = sigma
+    else:
+        wsigma = np.zeros(ydata.size)
+        if sigma is None:
+            for i in range(num_funcs):
+                wsigma[idxs[i]] = 1 / np.sqrt(weights[i])
+        else:
+            for i in range(num_funcs):
+                wsigma[idxs[i]] = sigma[idxs[i]] / np.sqrt(weights[i])
+
+    # Define multi-objective function
+    def f(x, *params):
+        y = np.zeros(x.size)
+        for i in range(num_funcs):
+            xi = x[idxs[i]]
+            yi = funcs[i](xi, *params)
+            y[idxs[i]] = yi
+        return y
+
+    # Run linearized curve_fit
+    analysis_result = curve_fit(f, xdata1d, ydata, p0, sigma=wsigma, **kwargs)
+
+    return analysis_result
+
+
+def curve_fit_data(
+    data: List[Dict[str, any]], data_processor: Callable, x_key: str = "xval", **filters
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Return array of (x, y, sigma) data for curve fitting.
+
+    Args
+        data: list of circuit data dictionaries containing counts.
+        data_processor: callable for processing data to y, sigma
+        x_key: key for extracting xdata value from metadata (Default: "xval").
+        filters: additional kwargs to filter metadata on.
+
+    Returns:
+        tuple: ``(x, y, sigma)`` tuple of arrays of x-values,
+               y-values, and standard deviations of y-values.
+
+    Raises:
+        QiskitError: if input data is not level-2 measurement.
+    """
+    filtered_data = filter_data(data, **filters)
+    size = len(filtered_data)
+    xdata = np.zeros(size, dtype=int)
```

Comment:
why integer?

Reply:
I think this is a typo from earlier version